torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -658,7 +658,7 @@ def _shard_parameters_(self, params_to_shard) -> None:
       self.sharded_params.append(p_shard)
       # Free the full parameter storage (here we free its `.data`) but keep the tensor itself
       # for auto-grad tracing (like `torch.autograd.Variable` before the tensor-variable merge).
-      p.data = p.data.new_zeros(1)
+      p.copy_(p.data.new_zeros(1))
```

Comment:
Is there a way for us to validate the data is still freed as expected?

Reply:
there is a counter corresponding to release of buffer.. let me look.

Reply:
for xrt https://github.com/pytorch/xla/blob/master/third_party/xla_client/xrt_computation_client.cc#L1285 should be called, I think we have similar function for PJRT.

Reply:
Let me double check. Thanks, Jack.

Reply:
Hit a crash while trying to replace all `.copy_` with `.set_` and then discovered that we need to lower `- set_.source_Tensor`. Trying to lower that op and see if it works.

Reply:
After commenting out the eval part, things can continue without any problems...

Reply:
Hmm, this is typically because some hooks didn't fire -- let me take a closer look at it

Reply:
Thanks, Ronghang. Yea, I'm not sure why would eval overwrites hooks.

Reply:
Hi @alanwaketan, I wonder could you verify whether the issue still occurs if we put the test loop in this line (https://github.com/pytorch/xla/blob/3f091bdd43944ee23a6acf17ef592f73adb06673/test/test_train_mp_imagenet_fsdp.py#L356) under `torch.no_grad`? (This shouldn't be a requirement for the test loop, but could help debug what's wrong.) If my guess is correct, then the error above should disappear when testing under `torch.no_grad`.

My guess is that during evaluation, the `_pre_backward_hook` somehow fired unexpectedly, causing the model to go into `TrainingState.BACKWARD_PRE` state. This hook needs to be registered during the forward pass unless it's under `torch.no_grad` (since we don't know if this is in an evaluation loop and there is a chance that the user would later call `loss.backward` even when the model is in eval mode). But registering this hook alone doesn't cause it to be tiggered, and the previous (expected) behavior is that this look should only fire when the user actually call something like `loss.backward()` -- but somehow this behavior is broken.

Reply:
@ronghanghu You are amazing... It solves the problem...
```
PJRT_DEVICE=TPU python3 test/test_train_mp_mnist_fsdp_with_ckpt.py --batch_size 16 --drop_last --num_epochs 2 --use_nested_fsdp
...
Epoch 2 test end 20:16:26, Accuracy=98.94
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000000-of-00000004.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000002-of-00000004.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000001-of-00000004.pth
checkpoint saved to /tmp/mnist-fsdp/final_ckpt_rank-00000003-of-00000004.pth
found 4 checkpoint files in /tmp/mnist-fsdp/final_ckpt_rank-*-of-*.pth
saved consolidated model to /tmp/mnist-fsdp/final_ckpt_consolidated.pth
Checkpoint consolidated, Accuracy=98.97 (note: it can be slightly different from the final training accuracy due to non-sync BatchNorm2d in the model)
Max Accuracy: 98.94%
```

Reply:
I see, thanks for verifying it quickly! Then it's weird why `_pre_backward_hook` would fire during a test loop when we don't call `.backward()` -- the expected behavior is that this hook wouldn't fire in such cases

Reply:
@alanwaketan I see, thanks for verifying it quickly! Then it's weird why _pre_backward_hook would now fire during a test loop when we don't call .backward() -- the expected behavior is that this hook wouldn't fire in such cases.

Reply:
On the other hand, I have observed that the number of `_post_backward_hook` being fired is incorrect for epoch 2. Maybe, without torch.no_grad, it actually modifies some other aspects of the model.

Reply:
@alanwaketan When looking at https://github.com/pytorch/xla/pull/4670, I suspect the issue there might be similar to what we had here -- somehow the `_pre_backward_hook` fired unexpectedly when we only do forward pass without a backward pass (e.g. when a submodule didn't participate in backward, or when we are running eval).

Reply:
Thanks for the heads up. I added a TODO.

Reply:
Minor update: I suspect this change in `_pre_backward_hook` might be related to the recent upstream change in PyTorch: [Hooks registered on tensor to always run, even if they are the inputs to .grad() (#85849)](https://github.com/pytorch/pytorch/pull/85849), as mentioned in the [PT 2.0 release note](https://gist.github.com/drisspg/09547b9ab16c5e236c496f120547c330#hooks-registered-on-tensor-to-always-run-even-if-they-are-the-inputs-to-grad-85849)